core/commands/add.go

```diff
@@ -61,47 +94,92 @@ remains to be implemented.
 					return
 				}
 
-				_, err = addFile(n, file, outChan)
+				_, err = addFile(n, file, outChan, progress)
 				if err != nil {
 					return
 				}
 			}
 		}()
-
-		return outChan, nil
 	},
-	Marshalers: cmds.MarshalerMap{
-		cmds.Text: func(res cmds.Response) (io.Reader, error) {
-			outChan, ok := res.Output().(<-chan interface{})
-			if !ok {
-				return nil, u.ErrCast()
-			}
+	PostRun: func(res cmds.Response) {
+		outChan, ok := res.Output().(<-chan interface{})
+		if !ok {
+			res.SetError(u.ErrCast(), cmds.ErrNormal)
+			return
+		}
+		res.SetOutput(nil)
 
-			quiet, _, err := res.Request().Option("quiet").Bool()
-			if err != nil {
-				return nil, err
+		quiet, _, err := res.Request().Option("quiet").Bool()
+		if err != nil {
+			res.SetError(u.ErrCast(), cmds.ErrNormal)
+			return
+		}
+
+		size := int64(0)
+		s, found := res.Request().Values()["size"]
+		if found {
+			size = s.(int64)
+		}
+		showProgressBar := size >= progressBarMinSize
+
+		var bar *pb.ProgressBar
+		var terminalWidth int
+		if showProgressBar {
+			bar = pb.New64(size).SetUnits(pb.U_BYTES)
+			bar.ManualUpdate = true
+			bar.Start()
+
+			// the progress bar lib doesn't give us a way to get the width of the output,
+			// so as a hack we just use a callback to measure the output, then git rid of it
+			terminalWidth = 0
+			bar.Callback = func(line string) {
+				terminalWidth = len(line)
+				bar.Callback = nil
+				bar.Output = os.Stderr
+				log.Infof("terminal width: %v\n", terminalWidth)
 			}
+			bar.Update()
+		}
 
-			marshal := func(v interface{}) (io.Reader, error) {
-				obj, ok := v.(*AddedObject)
-				if !ok {
-					return nil, u.ErrCast()
-				}
+		lastFile := ""
+		var totalProgress, prevFiles, lastBytes int64
 
-				var buf bytes.Buffer
+		for out := range outChan {
+			output := out.(*AddedObject)
+			if len(output.Hash) > 0 {
+				if showProgressBar {
+					// clear progress bar line before we print "added x" output
+					fmt.Fprintf(os.Stderr, "\r%s\r", strings.Repeat(" ", terminalWidth))
+				}
 				if quiet {
-					buf.WriteString(fmt.Sprintf("%s\n", obj.Hash))
+					fmt.Printf("%s\n", output.Hash)
 				} else {
-					buf.WriteString(fmt.Sprintf("added %s %s\n", obj.Hash, obj.Name))
+					fmt.Printf("added %s %s\n", output.Hash, output.Name)
+				}
+
+			} else {
+				log.Debugf("add progress: %v %v\n", output.Name, output.Bytes)
+
+				if !showProgressBar {
+					continue
 				}
-				return &buf, nil
+
+				if len(lastFile) == 0 {
+					lastFile = output.Name
+				}
+				if output.Name != lastFile || output.Bytes < lastBytes {
+					prevFiles += lastBytes
+					lastFile = output.Name
+				}
+				lastBytes = output.Bytes
+				delta := prevFiles + lastBytes - totalProgress
+				totalProgress = bar.Add64(delta)
```

Comment:
it's going to be painful implementing manual progress per-command. anything to be generalized here? (i get it that the `outChan` now returns more events, which are being listened to here.